plugins/inputs/windows_event_log/wineventlog/wineventlog_test.go

```diff
@@ -108,6 +115,154 @@ func TestReadWithBothSources(t *testing.T) {
 	assert.NoError(t, elog.Close())
 }
 
+func TestReadGaps(t *testing.T) {
+	originalWinEventApi := winEventAPI
+	defer func() {
+		winEventAPI = originalWinEventApi
+	}()
+
+	t.Run("BasicGapReading", func(t *testing.T) {
+		rl := state.RangeList{
+			// Inverted range: 0-5, 10-15, 20-25, 30-inf
+			state.NewRange(5, 10),
+			state.NewRange(15, 20),
+			state.NewRange(25, 30),
+		}
+		elog, stateFileName := newTestEventLogWithState(t, NAME, LEVELS, rl)
+		mockAPI := NewMockWindowsEventAPI()
+		winEventAPI = mockAPI
+
+		mockAPI.AddMockEventsForQuery(createMockEventRecordsRange(0, 5))
+		mockAPI.AddMockEventsForQuery(createMockEventRecordsRange(10, 15))
+		mockAPI.AddMockEventsForQuery(createMockEventRecordsRange(20, 25))
+
+		elog.Init()
+
+		var records []logs.LogEvent
+		// SetOutput calls run as well hence the omission of elog.run()
+		elog.SetOutput(func(e logs.LogEvent) {
+			records = append(records, e)
+			e.Done()
+		})
+		time.Sleep(8 * time.Second)
+		elog.Stop()
+
+		expectedRecords := []int{
+			1, 2, 3, 4, 5, 11, 12, 13, 14, 15, 21, 22, 23, 24, 25,
+		}
+
+		assert.Empty(t, elog.gapsToRead, "Gaps should be cleared after reading")
+		assert.Len(t, records, len(expectedRecords), "Should return correct number of mock events")
+		assert.Len(t, mockAPI.QueryCalls, 3, "Should make three query calls")
+		assert.Equal(t, NAME, mockAPI.QueryCalls[0].Path, "Should query correct path")
+		assert.Contains(t, mockAPI.QueryCalls[0].Query, "EventRecordID &gt; 0", "Query should contain start range")
+		assert.Contains(t, mockAPI.QueryCalls[0].Query, "EventRecordID &lt;= 5", "Query should contain end range")
+		assert.Contains(t, mockAPI.QueryCalls[1].Query, "EventRecordID &gt; 10", "Query should contain start range")
+		assert.Contains(t, mockAPI.QueryCalls[1].Query, "EventRecordID &lt;= 15", "Query should contain end range")
+		assert.Contains(t, mockAPI.QueryCalls[2].Query, "EventRecordID &gt; 20", "Query should contain start range")
+		assert.Contains(t, mockAPI.QueryCalls[2].Query, "EventRecordID &lt;= 25", "Query should contain end range")
+		assert.Greater(t, len(mockAPI.CloseCalls), 0, "Should make close calls")
+
+		for i, record := range records {
+			assert.Contains(t, record.Message(), fmt.Sprintf("Event %d", expectedRecords[i]))
+		}
+
+		assertStateFileRange(t, stateFileName, state.RangeList{
+			state.NewRange(0, 30),
+		})
+	})
+	t.Run("ReadGapThenSubscribe", func(t *testing.T) {
+		rl := state.RangeList{
+			// Inverted range: 2-4
+			state.NewRange(0, 2),
+			state.NewRange(4, 5),
+		}
+		elog, stateFileName := newTestEventLogWithState(t, NAME, LEVELS, rl)
+		mockAPI := NewMockWindowsEventAPI()
+		winEventAPI = mockAPI
+
+		// This is per EvtHandle hence the necessity to break up these calls
+		// 0, 1, 4 were "sent" previously (should be skipped)
+		mockAPI.AddMockEventsForQuery(createMockEventRecords(0, 2, 5))
```

Comment:
nit: Think this is incorrect. Should be 1, 2, and 5 were previously sent and should be skipped.